fda/grid.py

```diff
@@ -935,3 +935,37 @@ def __getitem__(self, key):
 
         else:
             return self.copy(data_matrix=self.data_matrix[key])
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+
+        for i in inputs:
+            if isinstance(i, FDataGrid) and not numpy.all(i.sample_points ==
+                                                          self.sample_points):
+                return NotImplemented
+
+        new_inputs = [i.data_matrix if isinstance(i, FDataGrid)
+                      else i for i in inputs]
+
+        outputs = kwargs.pop('out', None)
+        if outputs:
+            new_outputs = [o.data_matrix if isinstance(o, FDataGrid)
+                           else o for o in outputs]
+            kwargs['out'] = tuple(new_outputs)
+        else:
+            new_outputs = (None,) * ufunc.nout
+
+        results = getattr(ufunc, method)(*new_inputs, **kwargs)
+        if results is NotImplemented:
+            return NotImplemented
+
+        if ufunc.nout == 1:
+            results = (results,)
+
+        results = tuple((result
+                         if output is None else output)
+                        for result, output in zip(results, new_outputs))
+
+        results = [FDataGrid(data_matrix=r, sample_points=self.sample_points) for r in results]
```

Comment:
I'm sorry for responding so late. I have not tested it exhaustively but it looks good, I think it will improve the API a lot.

About the mutability of the data matrix I have been thinking some ways to deal with it, but I don't know which of them is better or if there are better ways.

* Construct the interpolator on each call to evaluate, but some methods, for example `shift_registration`, make multiple calls to evaluate and this could slow down the process too much.
* Check with a hash for changes in the data matrix before the evaluation to reconstruct the interpolator if necessary.
* Make the matrix read-only and remove the flag during the call to `array_ufunc`. After the call reset the interpolator, whenever called or checking with a hash.

About change operations  I would keep the current implementations and call the corresponding ufunc only in cases we do not have implemented.

On the subject of obsolete functions I agree that some should be removed, although they could still be used for bases. However I think `FDataBasis` should support ufuncs too, discretizing and re-projecting to the base for the internal evaluation.